qcodes/config/config.py

```diff
@@ -121,26 +123,35 @@ def update_config(self):
         """
         config = copy.deepcopy(self.defaults)
         self.current_schema = copy.deepcopy(self.defaults_schema)
-
-        if os.path.isfile(self.home_file_name):
-            home_config = self.load_config(self.home_file_name)
-            config = update(config, home_config)
-            self.validate(config, self.current_schema,
-                          self.schema_home_file_name)
-
-        if os.path.isfile(self.env_file_name):
-            env_config = self.load_config(self.env_file_name)
-            config = update(config, env_config)
-            self.validate(config, self.current_schema,
-                          self.schema_env_file_name)
-
-        if os.path.isfile(self.cwd_file_name):
-            cwd_config = self.load_config(self.cwd_file_name)
-            config = update(config, cwd_config)
+        self.config_file_path = path or self.config_file_path
+        if self.config_file_path is not None:
+            config_file = os.path.join(self.config_file_path, self.config_file_name)
+            schema_file = config_file.replace(self.config_file_name,
+                                              self.schema_file_name)
+            path_config = self.load_config(config_file)
+            config = update(config, path_config)
             self.validate(config, self.current_schema,
-                          self.schema_cwd_file_name)
-
-        return config
+                          schema_file)
+        else:
+            if os.path.isfile(self.home_file_name):
+                home_config = self.load_config(self.home_file_name)
+                config = update(config, home_config)
+                self.validate(config, self.current_schema,
+                              self.schema_home_file_name)
+
+            if os.path.isfile(self.env_file_name):
+                env_config = self.load_config(self.env_file_name)
+                config = update(config, env_config)
+                self.validate(config, self.current_schema,
+                              self.schema_env_file_name)
+
+            if os.path.isfile(self.cwd_file_name):
+                cwd_config = self.load_config(self.cwd_file_name)
+                config = update(config, cwd_config)
+                self.validate(config, self.current_schema,
+                              self.schema_cwd_file_name)
+
+        self.current_config = config
 
```

Comment:
I suppose we still want this method to return the config?

Reply:
why?

Reply:
this depends on how people use it. Unfortunately, `docs/examples/Configuring_QCoDeS.ipynb` does not shed light on it... I leave this decision to you.

Reply:
The missing return is the reason that the tests fail. I think we need a pretty strong argument for breaking/changing behaviour that we explicitly test for. We don't have visibility on who uses the software and how, so the golden rule is that we don't suddenly change the API. 

But I agree that the old/current API is not very intuitive, so we might deprecate some functions in favour of new ones that act better.